src/platform/posix/posix_clock.c

```diff
@@ -19,13 +19,46 @@
 
 #ifndef NNG_USE_GETTIMEOFDAY
 
-// Use POSIX realtime stuff
+#ifdef __APPLE__
+#include <mach/mach.h>
+#include <mach/mach_time.h>
+#endif
+
+// Use POSIX realtime stuff                                                          
 nni_time
 nni_clock(void)
 {
 	struct timespec ts;
 	nni_time        msec;
 
+#ifdef __APPLE__
+
+	int has_clock_gettime = 0;
+
+#if defined(MAC_OS_X_VERSION_10_12) && __has_builtin(__builtin_available)
+	if (__builtin_available(macOS 10.12, *)) {
+		has_clock_gettime = 1;
+	}
+#endif
+
+	if (!has_clock_gettime) {
+		// we could make this `__thread static` and read it only once,                                 
+		// not sure it's worth it, since the check for "the first time"                                
+		// introduces potential cache misses and other ways of making this                             
+		// run only once are more involved.                                                            
+		mach_timebase_info_data_t time_base_info;
+
+		// mach_continuous_time() is a better option, but it's only                                    
+		// available since MacOS 10.12, for which we already use clock_gettime().                      
+		uint64_t absolute_time = mach_absolute_time();
```

Comment:
Not sure what you're aiming for here. The latest version is now complete, it can even build on older SDKs. Unless you have more stylistic remarks. Not sure how doing something completely anew would "expedite things'. 

Also, what is used with `clock_gettime()` can be changed. Currently we do use `CLOCK_REALTIME`, but not everybody does that, AFAICT by looking around for `nng` clones on Github. We probably should change that, but that is a separate matter.

OTOH, in this same file, there is a huge comment against using `gettimeofday()`:

> // If you're here, its because you don't have a modern clock_gettime with
> // monotonic clocks, or the necessary pthread_condattr_settclock().  In
> // this case, you should be advised that *bad* things can happen if your
> // system clock changes time while programs using this library are running.
> // (Basically, timeouts can take longer or shorter, leading to either hangs
> // or apparent spurious errors.

Reply:
What I'm saying is that based on how nng is currently implemented for apple, it calls clock_gettime() using CLOCK_REALTIME, but you implemented the fallback code as though it was called with CLOCK_MONOTONIC, which has two issues : A) its not meeting the contract of how it was called, and B) the monotonic implementation doesn't meet the contract defined for CLOCK_MONOTONIC.

So i'm saying, rather than try and figure out how to formulate a patch that fits in properly with how NNG works, lets just fix OUR problem so that it works the same on all versions of macOS as NNG is implemented today (which was the original scope of the work, which once that was done, we were to look at if it made sense to provide a patch upstream).

in the end, its best not to end up in a situation where we HAVE to maintain a fork.

Reply:
Then I guess something like #2 would be along those lines?